### DISCUSSION TITLE
No event driven MQTT messages??

### DISCUSSION THREAD

**Discussion by Moosbueffel** [arendst/Tasmota#11610] [Q&A] [answered]
I'm a little bit confused. I thought, that with Tasmota and MQTT I get an event driven system. Meaning, that I get a message every time - and only than - when a value changes. Perhaps it is possible but I didn't found a document/examples/hints

I'm using Tasmota with Sonoff e.g. a POW

What I found out is:
1) I can listen to a topic and get a message every 5 min at default.
2) I can reduce this time with TeleMetry
3) this work,  but heavy load my network and it's unnecessary most of time, because value didn't change
4) I can use http instead and poll in different times e.g day and night. But it's always polling
5) I found out, that the switch will work as I thought via the stat topic (stat/POW_01/POWER and stat/POW_01/RESULT)
6) Why is this message of POWER status double. As a String and As a object**?**
7) If I poll with http, I get also a stat message, so I get the result also double!?

So as my summary, I can get the status of the switch with the stat topic. Than I have to poll for Voltage, Current,... My current Node-Red will look like this:
![grafik](https://user-images.githubusercontent.com/7726457/113733926-2d8b6580-96fb-11eb-8ffc-295652f52a4b.png)

This is subotimal. But how can I get event driven messages, so I get a MQTT messeage every time a value changes - and only than!? I can't imagine, that only I want to get this!?

Thanks for every answer (for a newbie) or a link with the decription/example how to do it

And a question about TeleMetry: Are the values written to flash, because I can write to the flash not so often. So is it a bad idea to often change telemetry?

**Comment by nonix** [comment]
Hi, I don't know / don't have the POW but I would guess the power meter is sourced via A0 input. Perhaps you could try to trigger a rule on Analog change. 
The A0 would send values between 0 and 1024 based on analog read. This could produce a lot of msgs even if "no change" just by wire noise. To reduce noise you could try A0Div10 as this will trigger only when changed by +/- 1%.

The provided info is not verified to work with POW, take it with a grain of salt.

Cheers,
 N.

**Comment by barbudor** [comment, marked as answer]
@Moosbueffel 

Telemetry is send periodically. In itself it doesn't trigger any flash write.
If you want to get messages only for some changes, you can use Rules to detect the changes and publish on a specific custom topic. Then your NodeRed will pickup only this topic and not the tele//STATE nor tele//SENSOR messages
See: https://tasmota.github.io/docs/Rules/
and https://tasmota.github.io/docs/Rules/#transmit-sensor-value-only-when-a-delta-is-reached
and https://tasmota.github.io/docs/Rules/#publish-maximum-value-from-sensor-in-a-time-period
to give you ideas

Note that for Energy measurement, this function is already built-in as [`PowerDelta`](https://tasmota.github.io/docs/Commands/#powerdelta)

Other messages are indeed event triggered 

POWER are send both Text and JSON for historical reasons
If you set your MQTT Node output to "Parsed JSON Object", then Text messages will be ignored
You can also use [`SetOption90 1`](https://tasmota.github.io/docs/Commands/#setoption90) to disable Text version of the message

Let us know if you have further question

**Comment by Moosbueffel** [comment]
Many thanks @barbudor, for detailed description and links. But I'm not successful. I read all your links, but have always problems in understanding. But it seems to me a good thing to use the Rules, so I wnat to understand and use it.

Because your write, that a change in TelePeriod will not be written to flash I thougt this will be the easiest way for a first shot and if this work I can improve it:

1) I tried different things:
Rule1 ON POWER=1 DO TelePeriod 10 ENDON ON POWER#State=0 Do TelePeriod 0 ENDON
also with POWER==1, POWER.State=1, POWER.State==1, POWER=off

**Nothing worked!** So what is the correct syntax? Or can't I use stat and have to use Sensor data instead (energy.power>1)? 

**This work:**
`Backlog Rule1 ON ENERGY#Voltage>1 DO TelePeriod 10 ENDON ON ENERGY#Voltage<1 DO ruletimer1 12 ENDON ON rules#timer=1 Do TelePeriod 0 ENDON; Rule 5; Rule1 1`

2) I tried to code my target in a Rule but I'm moving in a circle because of missing if, than, else and/or bit operations. I tried something like this:

`Rule1 
ON ENERGY#Voltage=0 DO var1 0 ENDDO 
ON ENERGY#Voltage>0 DO var1 1 ENDDO
ON ENERGY#Current=0 DO ???? to reset current setting at decimal 10
ON ENERGY#Current>0.5 DO ADD1 10 ENDDO
publish %var1% ????

Rule 5
Rule1 1`

In pseudo code, what I wanted to do:

`
ON voltage=off DO publish off 
ON voltage=on  DO if current>0 publish heating
				  if current=0 publish hot
ON current=0 DO oublish hot
`
The last case (on current=0) is, if the heating is switched of from boiler side, so voltage is on, but the current=0. Background is to use my photovoltaic power and visualize it.

Is there a better solution?


3) Is there a way to send an immediate teleperiod message. Setting teleperiod 0 will suppress the sending of  "power off", so I did the trick with the timer.

4) Is a Rule store in flash and provided after power off/on? Or does I have to set savedate 1? abfter the rule? Where is a description of what is stored in flash and perhaps where I have to be careful writing values, because they are stored in flash?

**Reply by sfromis** [reply]
Note that the telemetry updates published at TelePeriod intervals does indeed not cause flash writes, but updating the TelePeriod setting as such **will** cause flash writes, unless you've turned that off. 

A key thing about how rule triggers work, is that it is **not** to test current value, but to react to value being updated (changed or not). If you want to react to actual changes, you can have a trigger test if it is different, and then save the new value for next time.

**Reply by Moosbueffel** [reply]
> Note that the telemetry updates published at TelePeriod intervals does indeed not cause flash writes, but updating the TelePeriod setting as such **will** cause flash writes, unless you've turned that off.

And how can I do turning off TelePeriod changes writes to flash? 

So can I trigger on change itself like 
ON ENERGY#Current DO ... ENDDO

**Reply by sfromis** [reply]
Using `SaveData` you can turn off all automatic writing of settings to flash. But varying TelePeriod is likely not a good strategy anyway. If you "just" want to publish telemetry now, you can just issue the TelePeriod command without arguments. Will publish telemetry without flash write. 

But it is not clear what the aim of your efforts is at all. It may be doable without having to do guesses backtracking trough the things you're trying to do.

**Comment by barbudor** [comment]
Why do you want to change Teleperiod ?
As I said, Teleperiod occurs regularly, publish sensors and information to MQTT, all those without writing to flash
You CAN'T DISABLE Teleperiod. `Teleperiod 0` has no meaning, the minimum is 10 seconds.

So can you please elaborate on your use case as we don't understand what exactly you want to achieve.

**Reply by Moosbueffel** [reply]
> You CAN'T DISABLE Teleperiod. `Teleperiod 0` has no meaning

Really? I read this at the [documents](https://tasmota.github.io/docs/Commands/):
`

TelePeriod | See current value and force publish STATE message
0 = disable telemetry messages
`

**Reply by barbudor** [reply]
My bad, you are right. I was so sure.

**Comment by Moosbueffel** [comment]
Many thanks @barbudor and @sfromis for your patience. And sorry for my bad english explaining my usecase/thoughts. But perhaps I have a basically missunderstanding. So I go back to explain what I want to do.

**usecase**
As I tried to explain above I have a boiler (water heater), which I want to control in dependence from the power I get from my photovoltaic. The reason is to rise the amount of self consumption. The switch is a Sonoff POW with a switch and a voltage, current measurement. In this scenario I have this states:

1) no photovoltaic power ==> heater off
2) good photovoltaic power ==> heater on
    In case of heater on I have this two substates:
             a) heater finished (current == 0)
             b) heater heats (current > 0)

**what I want to do**
The switch (boiler on/off) is controlled in dependence of measured photovoltaic power. It's working.

But I also want to get the oeration mode of the boiler for further controlling and visualisation. For this I have only the status messages of the POW (POWER, voltage, current). With this I want to differentiate the following modes:
- heater off
- heater on and heating
- heater on an finished

In case of "heater off" I have nothing to do. And I don't need messages from tasmota.
In case of "heater on" I need messages in short interval to differentiate case heating and finished.

** first idea**
What I missed at the beginning was an event triggered message from tasmota (see the Thread title and first post). My idea was that I get a "tele" every time, a value of POW changes. In this case, I
- didn't get any messages, as long as heater is off
- get a message, when power changes to on ==> case 2a or 2b ==> default case 2a (heater finished9
- get a message when current changes from 0 to x ==> case 2b (heater heats)
- get a message when current changes from x to 0 ==> case 2a heater finished
- get a message when power off ==> case 1 (not enough photovoltaic for water heater)

In this event triggered scenario, I get only messages when something happens and I have e.g. to update my visualisation, displaying the obove mentioned states.

**what I tried**
can you read in the above posts. 
- First was the realisation in Node-Red with a timer and http requests in case of power on to notice change of current (I can post code, if interesting). See picture in my first post
 - Second was to change the teleperiod. 
         - Power off = teleperiod 0 (no messages)
         - Poer on = teleperion 10 ( to notice current changes)

this result in a shorter Node-Red:
![grafik](https://user-images.githubusercontent.com/7726457/114375197-b9324580-9b84-11eb-8f4a-047726bba252.png)

The code of the function node:
`
var voltage = msg.payload.ENERGY.Voltage;
var current = msg.payload.ENERGY.Current;
var label, color;

if(voltage === 0){
    label = "Boiler aus";
    color = "black";
} else {
    if(current === 0) {
        label = "Boiler an (fertig)";  
        color = "grey"
    } else {
        label = "Boiler an (heizt)";
        color = "red";
    }
}

msg.label = label;
msg.color = color;
// msg.tooltip =  "Heute "+msg.payload.StatusSNS.ENERGY.Today+" kWh";
msg.tooltip =  "Heute "+msg.payload.ENERGY.Today+" kWh";


return msg;
`

**Summary**
The goal in this and other projects is to get MQTT messages only, if I really need them. e.g. in case of POWER==off I don't need any MQTT messages. In another application I have nearly the same scenario: I don't need messages till the whole day. Only when I'm starting to watch television at night, I need messages. And than I need the messages - best event triggered - or in short periods, so I can react immediately.

**Perhaps I have a basic missunderstanding??** How do you handle such things?

**Reply by barbudor** [reply]
Hi @Moosbueffel 
I understand but why would you like to stop Tasmota from sending messages. It's a useless optimization.
I have a few [power plugs](https://templates.blakadder.com/gosund_SP111_v1_1.html) with energy monitoring that I am using for various use case.
I generally have the default Teleperiod of 5min + PowerDelta to report changes for more than 15-20%
I don't care of getting messages even when I don't need them. 
NodeRed neither care.

Your state machine can just ignore those that you don't need like when POWER is off.
I think you are scratching your head for little benefice.

**Comment by Moosbueffel** [comment]
The reason is, that I want to synchronice production and consumption and of course control it. That's for increasing the power I get from my photovoltaic, because I have to pay 28ct/kWh but get only 9,8ct/kWh for my produced power. So my intention is to use as much as possible from my own. 

To do this and for synchron display of production and consumption I need a short teleperiod. But there are cases this isn't necessery to report, as I described bevor. My intention now was to reduce network traffic, powerconsumtion of Tasmota device and Rasperry (perhaps it's not so much) and to reduce CPU power for this on my Raspberry which has to do many other things.

But perhaps you are right and I can ignore the first two points. For raspberry CPU power  it will be a problem in future, when the system is growing. So my try was to get only messages, when somethig happend. And I thougt polling - or regular sending of messages - is something from the past ;-).

**Reply by sfromis** [reply]
For energy monitoring data, you do not need a short TelePeriod. Use PowerDelta as already suggested by barbudor.

**Reply by Moosbueffel** [reply]
I invest some hours today to try the PowerDelta. This is the worst solution in my mind.

The problem is, that there is no event when changing from off to on. So I have to implement the stat/POW_01/POWER to get an event in case of power switches to on. Than I have to send a http request to get current data of volatage and current. 

Second I also need this in case of power switches to off. I get a SATUS8 in this case with  energy data, but it's the old date (at the time, the switch was on) and not the current state, (switch is off). So in this case I set the wrong output and have to override it after an additional http request.

So my conclusion is:
-  that the powerDelta is for another use case.
- best solution would be changing the teleperiod. It's well to understand and the shortest node red flow. Disadvantage only the rule in tasmota device.
- second best will be the first solution at this thread. Advantage is not to need a rule at tasmota and no need to change anything. Disadvantage a more elaborate process in node red flow.

Add the end I learned a lot. So have many thanks to all distributors

**Reply by sfromis** [reply]
No, you do not need http requests for **anything**. Using PowerDelta, you can get a message when the energy level switches "enough", which will also be when turning off/on. Changing TelePeriod is not a good way, when you can simply get messages automatically when things happens. You may have to create a bit of logic to detect if PowerDelta is because of going to or from zero.

**Reply by Moosbueffel** [reply]
Hmm - I tried a powerDelta of 20 and a powerDelta of 120 (relativ and absolut) and get the discribed behavior. Whereby the value doesn't matter.  I have 0 or 450W.

I give it an additional try tomorrow, but don't know, what could be wrong. Perhaps I will try a really new setup.

**Reply by Moosbueffel** [reply]
Ok, I tested it at another device. There it works :-). So I have to look at the original device, whats wrong. I have there an oder tasmota V7.1.2 perhaps that is the reason

**Reply by sfromis** [reply]
In 8.1.0.4, an issue about "PowerDelta zero power detection" was fixed. Not sure of any details, or if it makes a difference for you, but perhaps.....  #7515

**Reply by Moosbueffel** [reply]
Fantastic! After Update to current release PowerDelta works :-). And also my flow. Changing the teleperiod isn't needed any more :-). Perfect!'

So many thanks for help and patience!
